hack/generator/pkg/codegen/pipeline_flatten_properties.go

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/azure-service-operator/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+)
+
+func flattenProperties() PipelineStage {
+	return MakePipelineStage("flatten-properties", "Apply flattening to properties marked for flattening", applyPropertyFlattening)
+}
+
+func applyPropertyFlattening(
+	ctx context.Context,
+	defs astmodel.Types) (astmodel.Types, error) {
+	visitor := makeFlatteningVisitor(defs)
+
+	result := make(astmodel.Types)
+	for name, def := range defs {
+		newDef, err := visitor.VisitDefinition(def, name)
+		if err != nil {
+			return nil, err
+		}
+
+		result.Add(newDef)
+	}
+
+	return result, nil
+}
+
+func makeFlatteningVisitor(defs astmodel.Types) astmodel.TypeVisitor {
+	return astmodel.TypeVisitorBuilder{
+		VisitObjectType: func(this *astmodel.TypeVisitor, it *astmodel.ObjectType, ctx interface{}) (astmodel.Type, error) {
+			newIt, err := astmodel.IdentityVisitOfObjectType(this, it, ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			it = newIt.(*astmodel.ObjectType)
+
+			newProps, err := flattenObjectType(it, defs)
+			if err != nil {
+				return nil, err
+			}
+
+			// safety check:
+			if err := checkForDuplicateNames(newProps); err != nil {
+				return nil, err
+			}
+
+			result := it.WithoutProperties().WithProperties(newProps...)
+
+			return result, nil
+		},
```

Comment:
I really dislike putting methods inline like this - the code becomes impenetrable. Recommend pulling out the anonymous functions as methods, giving them an intention revealing name.

Reply:
In this case I don't think there are much better names to use: the functions are very specific to the flattening visitor so it would just be something like `visitObjectTypeAndFlatten` which isn't any better than the `VisitObjectType:` field inside the `flatteningVisitor` method… 🤔

Reply:
Meh. I still think having them inline makes the method too long to keep in your head as it's read, but it's not a blocker.